### DISCUSSION TITLE
Discussing Proper Polymod Support

### DISCUSSION THREAD

**Discussion by isakube** [KadeArchive/Kade-Engine#1319] [Ideas]
# This discussion will be related to integrating Polymod into Friday Night Funkin': Kade Engine the intended way.

## Overview
If you go at most 30 lines of code through the current master source's `TitleState.hx` (not counting the shitload of imports,) you will find a polymod initialization. This initialization, however, is not set up correctly. If you copy the `getIntroTextShit()` function, change the text source to `mods/modList.txt`, and remove the `for()` loop at the end (as well as update the returned value type to a String Array,) you can call the new function in the `dirs:` part of the polymod initalization. AND IT WORKS! Well, not entirely.

## Probably the Problem
I believe the reason Polymod doesn't work has something to do with our [caching system.](https://github.com/KadeDev/Kade-Engine/blob/stable/source/Caching.hx) Since initializing Polymod clears the asset cache, some assets may not be loaded properly. "Some assets" includes, but is not limited to, `webm`, `png`, and `ogg` files. So pretty much everything except text.

The reason only text files can be replaced is because:

### Short:
We don't cache text, therefore it can be replaced.

### Long:
Our caching process only caches media. (Which, in this case, means anything but text.) The whole purpose of the caching process is to use it, so instead of utilizing the Path.hx functions, (Or maybe we do. I haven't looked.) we grab the asset from KE's cache. Because of where we initialize Polymod (TitleState), the modified files that Polymod loads don't get cached. Instead, the vanilla files get cached. Therefore, using Polymod to modify images or music doesn't work.

### But wait..
If initializing Polymod clears the asset cache, how come the vanilla files still display? Shouldn't they be missing?
No. Polymod doesn't work with HaxeFlixel, the library we use to cache assets.

## Solution?
1. Drag the [aforementioned modifications](#overview) into Caching.hx, and make sure the Polymod initialization is at the top of the `create()` function.
2. Use Lime / OpenFL to cache assets instead of HaxeFlixel.
### Issues where Polymod support is mentioned (and would be useful in helping create/fix): 
(feel free to add more in the comments)
- #1302 

###### I spent a long time formatting this. I hate being formal but I do like Markdown.

**Comment by EliteMasterEric** [comment]
Polymod notably provides the infrastructure needed for multiple mods to co-exist without overlapping or interfering with each other, and without requiring special modification of the source code. It also provides the backend infrastructure needed for stuff like a mod menu and API versions, allowing for stuff like custom songs, custom stages, custom weeks, custom note types, and HScript support (far more powerful than Lua modcharting).

My perfect, ideal implementation, would be flexible and expansive enough to allow something like the Tricky mod to be implemented using only Polymod.

The base game includes a stub of a Polymod implementation that barely works (hence why Polymod is a dependency), but new PRs to Polymod itself have provided the fixes needed for it to be used within OpenFL.

I am working on a PR currently that tries to implement this. I have a base written but am still working on getting image assets loaded.

**Reply by EliteMasterEric** [reply]
![image](https://user-images.githubusercontent.com/4635334/133393710-3f6c1b78-f304-4403-b809-7693a6f1f970.png)

Just going to leave this here.

**Comment by EliteMasterEric** [comment]
The one concern is that I have seen Kade mention that this engine is focused on developers who plan to build it from source, and therefore rearranging data to be more accommodating of mod support like this is less of a priority, and may even be rejected outright.

**Reply by isakube** [reply]
yeah.
your PRs could be fitted into current vanilla Funkin' source, and it would make more sense there.

**Reply by EliteMasterEric** [reply]
@prokube The changes to the Polymod library that made this possible were literally made at ninjamuffin99's request, according to the Polymod dev.

A system similar to this IS what proper mod support for FNF will look like. I'm simply jumping the gun a bit by implementing it into the most popular engine for modding, and I'm finding and fixing a lot of pain points in the process.
